cranelift/wasm/src/code_translator.rs

```diff
@@ -3579,50 +3581,260 @@ where
     iter_ptr: Value,
+    starting_offset: i64,
     tagged_ptr: Value,
     size: i64,
     builder: &mut FunctionBuilder,
-) -> WasmResult<()> {
+) {
+    let mut i = starting_offset;
+    while size - i >= 32 {
+        let cur_ptr = builder.ins().iadd_imm(iter_ptr, i);
+        builder.ins().arm64_st2g(tagged_ptr, cur_ptr);
+        i += 32;
+    }
+    if i < size {
+        let cur_ptr = builder.ins().iadd_imm(iter_ptr, i);
+        builder.ins().arm64_stg(tagged_ptr, cur_ptr);
+    }
+}
+
+/// Tag a memory region which has a static size known at compile-time.
+/// Optimizations include:
+/// - Use st2g operation to tag two memory regions at once
+/// - Only unroll loops until a certain threshhold, since very large sizes
+///   increase number of generated instructions and increase cache misses
+fn tag_memory_region_static<FE>(
+    iter_ptr: Value,
+    tagged_ptr: Value,
+    size: i64,
+    size_dynamic: Value,
+    builder: &mut FunctionBuilder,
+    environ: &mut FE,
+) -> WasmResult<()>
+where
+    FE: FuncEnvironment + ?Sized,
+{
+    // Runtime code that is generated:
+    //
+    // (compile-time) num_required_st2g_loops = floor(size / LOOP_UNROLL_BYTES_THRESHHOLD)
+    // (compile-time) num_required_last_stg = (size % LOOP_UNROLL_BYTES_THRESHHOLD) / 16
+    //
+    // (compile-time) if num_required_st2g_loops <= 1
+    // (compile-time)     generate WITHOUT_LOOP_UNROLLING_THRESHHOLD
+    // (compile-time) else
+    // (compile-time)     generate WITH_LOOP_UNROLLING_THRESHHOLD
+    //
+    // WITHOUT_LOOP_UNROLLING_THRESHHOLD:
+    // +---less_than_threshhold
+    // | st2g(tagged_ptr, iter_ptr + 0)          \
+    // | st2g(tagged_ptr, iter_ptr + 32)          | repeat `n` times, i.e. while (remaining_bits >= 32) holds
+    // | ...                                      |
+    // | st2g(tagged_ptr, iter_ptr + (n-1) * 32) /
+    // |
+    // | stg(tagged_ptr, iter_ptr + n * 32)  | generate only if required
+    // +---
+    //
+    // WITH_LOOP_UNROLLING_THRESHHOLD:
+    // st2g_loop_counter = 0
+    // +---st2g_loop_condition_block(iter_ptr, st2g_loop_counter)
+    // | if (st2g_loop_counter < num_required_full_st2g_loops)
+    // |     goto st2g_loop_body_block(iter_ptr, st2g_loop_counter)
+    // | else
+    // |     (compile-time) if (num_required_last_stgs == 0)
+    // |                        goto next_block
+    // |     (compile-time) else
+    // |                        goto last_stgs_block
+    // |
+    // | +---st2g_loop_body_block(iter_ptr, st2g_loop_counter)
+    // | | st2g(tagged_ptr, iter_ptr + 0)          \
+    // | | st2g(tagged_ptr, iter_ptr + 32)          | repeat `n`=LOOP_UNROLL_BYTES_THRESHHOLD times
+    // | | ...                                      |
+    // | | st2g(tagged_ptr, iter_ptr + (n-1) * 32) /
+    // | |
+    // | | iter_ptr += LOOP_UNROLL_BYTES_THRESHHOLD
+    // | | st2g_loop_counter++
+    // | | goto st2g_loop_condition_block(iter_ptr, st2g_loop_counter)
+    // | +---
+    // +---
```

Comment:
I actually think it would make sense to just use the `iter_ptr`, which we have to increment every st2g-loop anyways, instead of `st2g_loop_counter`. Furthermore, we are calculating `end_ptr = original_iter_ptr + size` once at the beginning anyways to check if it overflows, so we might as well replace the condition `if (st2g_loop_counter < num_required_full_st2g_loops)` with `if (iter_ptr < end_ptr)` and save the instructions involving `st2g_loop_counter++`. Let me know what you think.